lib/CL/devices/level0/level0-compilation.hh

```diff
@@ -0,0 +1,257 @@
+class Level0Program {
+  // all data except Builds, Kernels & BuildLog are const
+  std::mutex Mutex;
+  // vectors of multiple builds with different set of build settings,
+  // shared_ptr because outsiders can hold a build instance & wait for it to finish
+  std::list<Level0ProgramBuildUPtr> Builds;
+  std::list<Level0KernelUPtr> Kernels;
+  std::string BuildLog;
+
+  ze_context_handle_t ContextH;
+  ze_device_handle_t DeviceH;
+  std::string CacheDir;
+  std::string CacheUUID;
+
+  // SPIR-V binary data
+  std::vector<uint8_t> SPIRV;
+  // spec constants used for builds
+  ze_module_constants_t SpecConstants;
+  // storage for actual data of ze_module_constants_t
+  std::vector<uint32_t> ConstantIds;
+  std::vector<const void*> ConstantVoidPtrs;
+  std::vector<std::vector<uint8_t>> ConstantValues;
+
+
+  void setupSpecConsts(uint32_t NumSpecs, uint32_t* SpecIDs,
+                      const void **SpecValues, size_t *SpecValSizes);
+
+public:
+  Level0Program(ze_context_handle_t Ctx,
+                ze_device_handle_t Dev,
+                uint32_t NumSpecs, uint32_t *SpecIDs,
+                const void **SpecValues, size_t *SpecValSizes,
+                std::vector<uint8_t> &SpvData,
+                const char* CDir,
+                const std::string &UUID);
+  ~Level0Program();
+
+  void addFinishedBuild(Level0ProgramBuildUPtr Build);
+
+  std::vector<uint8_t>& getSPIRV() { return SPIRV; }
+  ze_module_constants_t getSpecConstants() { return SpecConstants; }
+  ze_device_handle_t getDevice() { return DeviceH; }
+  ze_module_handle_t getAnyHandle(); // for setup_metadata
+  Level0Kernel *createKernel(const char* Name);
+  bool getBestKernel(Level0Kernel *Kernel, bool LargeOffset,
+                        ze_module_handle_t &Mod, ze_kernel_handle_t &Ker);
+  bool releaseKernel(Level0Kernel *Kernel);
+  std::string &&getBuildLog() { return std::move(BuildLog); }
+  const std::string &getCacheDir() { return CacheDir; }
+  const std::string &getCacheUUID() { return CacheUUID; }
+};
+
+typedef std::shared_ptr<Level0Program> Level0ProgramSPtr;
+
+class Level0CompilationJob {
+  bool HighPrio;
+  bool Finished;
+  bool Successful;
+  Level0ProgramSPtr Program;
+  Level0ProgramBuildUPtr Build;
+
+  std::mutex Mutex;
+  std::condition_variable Cond;
+
+public:
+  Level0CompilationJob(bool HiP, Level0ProgramSPtr P, Level0ProgramBuildUPtr PB)
+    : HighPrio(HiP), Finished(false), Successful(false),
+      Program(P), Build(std::move(PB)) {}
+  ~Level0CompilationJob() = default;
+  bool isHighPrio() { return HighPrio; }
+  ze_device_handle_t getDevice() { return Program->getDevice(); }
+  Level0ProgramBuild* getBuild() { return Build.get(); }
+  bool isForProgram(Level0Program *P) { return Program.get() == P; }
+  void signalFinished();
+  bool isSuccessful() { return Successful; }
+  void waitForFinish();
+};
+
+typedef std::shared_ptr<Level0CompilationJob> Level0CompilationJobSPtr;
+
+/* A queue of compilation jobs with some extra features,
```

Comment:
Use C++ comments as advised by the LLVM C++ coding guidelines.

Reply:
fixed